### ISSUE TITLE
Newsletter subscription does not set the correct store_id if already subscribed.  Not Fixed in 2.3-dev

### ISSUE THREAD

**Issue by bbmxxc** [magento/magento2#19172]
Regarding https://github.com/magento/magento2/issues/19131.
I (again) tried it on 2.3-dev an the issue is *not* fixed in 2.3-dev.

### Preconditions (*)

1. latest Magento Version 2.3-dev
2. Have at least two stores ( A and B)
### Steps to reproduce (*)

1. Register a user in  A
2. Subscribe to Newsletter in A 
3. Switch to Store B
4. Subscribe to Newsletter again

### Expected result (*)

1. When in store B, Subscriber::loadByEmail() should work
2.` SELECT store_id, customer_id FROM newsletter_subscriber` should look like this:

```
+----------+-------------+
| store_id | customer_id |
+----------+-------------+
|        1 |           1 |
|        2 |           1 |
```
 

### Actual result (*)

1. When in store B, loadByEmail() (or loadById() ) cant find the subscriber since
2. `SELECT store_id, customer_id FROM newsletter_subscriber` looks like this:
```
+----------+-------------+
| store_id | customer_id |
+----------+-------------+
|        1 |           1 |
|        1 |           1 |

```

### What i think where the error is
In `app/code/Magento/Newsletter/Model/Subscriber.php` starting on line [468](https://github.com/magento/magento2/blob/6481cecaa9ea6959ebb255b4271f88d746ebb21b/app/code/Magento/Newsletter/Model/Subscriber.php#L468): 
If the subscriber already exists with this email, the new entry gets the `$customer->getStoreId()`
not the `$this->_storeManager->getStore()->getId()`

If this is the intended behavior then `loadByEmail()` or `loadByCustomerId()` should be store_id agnostic.

Cheers!
Ben

**Comment by magento-engcom-team** [comment]
Hi @bbmxxc. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@bbmxxc do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [x] yes
- [ ] no

**Comment by ghost** [comment]
Hi @bbmxxc on fresh magento 2.3-dev instance 
![selection_199](https://user-images.githubusercontent.com/40756439/48353488-34336100-e698-11e8-8c92-607401549be4.png)

**Comment by bbmxxc** [comment]
Hi! Thanks for the fast reply. But this are all different users. Try registering an user in one store and subscribing to the newsletter in a different one.

**Comment by ghost** [comment]
@bbmxxc but if i switch to store B then user logout.

**Comment by bbmxxc** [comment]
Well my customer stays logged in. But you could still log him in again at the other store.

**Comment by bbmxxc** [comment]
@magento-engcom-team give me $VERSION instance

**Comment by bbmxxc** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @bbmxxc. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @bbmxxc, here is your [Magento instance](http://34.228.235.121/i-19172-2-3-develop/).
Admin access: [http://34.228.235.121/i-19172-2-3-develop//admin](http://34.228.235.121/i-19172-2-3-develop//admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by bbmxxc** [comment]
I rebuild the situation on requested instance. There is StoreA and StoreB and a user registered on StoreA. I have subscribed to the newsletter on both stores but the store_id saved to the subscription is always that of StoreA.

**Comment by magento-engcom-team** [comment]
Hi @eduard13. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by eduard13** [comment]
Hi @bbmxxc I'm not sure how could you get the following result:
```
+----------+-------------+
| store_id | customer_id |
+----------+-------------+
|        1 |           1 |
|        1 |           1 |
```
As there is no way to subscribe the customer, because he's already subscribed in another store. Can you give me more details regarding this?
I followed the same steps, but couldn't get the duplication rows in DB (tried in 2.2 and 2.3).

I agree that there is an issue regarding the customer subscription from different stores and I'm going to investigate this.
The steps to reproduce:
1. Create 2 stores: **A**(default) and **B**
2. Login on the frontend
3. Subscribe to Newsletter (which should be successfully)
4. Switch to store B
5. Subscribe to Newsletter (you get the success message, but the subscription wasn't saved)

Thank you.

**Comment by magento-engcom-team** [comment]
@eduard13 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-96324`, `MAGETWO-96325` were created

**Comment by bbmxxc** [comment]
Hey @eduard13!

I for subscribing in store **B** i use the subscription box in the lower right corner. Or Programmatically by calling `Magento\Newsletter\Model\Subscriber::subscribe(...)`.

**Comment by eduard13** [comment]
@bbmxxc thank you for informing, I will take it into account while investigating this.
Thank you.

**Comment by wojtekn** [comment]
I also spot issue related to this and there is another way to reproduce it, in backend:

1. Install clean Magento 2.2.x or 2.3.x (sample data not needed)
2. Log in into admin backend
3. Create a second website, store and store view
4. Open the second website frontend
5. Register new customer account and subscribe to the newsletter during registration
6. Back to the backend
7. Edit customer, open Newsletter tab. Note that he is not marked as subscribed
8. Select newsletter checkbox to sign him up
9. Click "Save and Continue Edit"

A success message is displayed, the customer is added to the newsletter_subscriber table, but the customer is still not marked as subscribed in the edit form. Doing steps 7 to 9 results in duplicating entry in subscribers table. After few saves I see following:

    +---------------+----------+-------------+-------------------+
    | subscriber_id | store_id | customer_id | subscriber_status |
    +---------------+----------+-------------+-------------------+
    |             3 |        2 |           3 |                 1 |
    |             4 |        2 |           3 |                 1 |
    |             5 |        2 |           3 |                 1 |
    +---------------+----------+-------------+-------------------+

Additionally, it sends an email confirmation to the customer each time.

The same problem occurs when such customer is subscribed or saved by CLI or CRON job.

It looks it was introduced in the https://github.com/magento/magento2/commit/65000d81b388a7218ae9c9ec5c12bcb00d932c3c commit.

The problem doesn't occur in 2.1.5 as `loadByCustomerData` method was ignoring store_id there. 

Maybe `\Magento\Newsletter\Model\Subscriber::loadByCustomerId` shouldn't overwrtite store_id on customer record? In my case it seems to fix the problem.

Another issue is that when customer for the new website is created by admin in the backend and he is subscribed to the newsletter by admin, store_id  of his record in newsletter_subscriber is 1. This can lead to other buggy use cases.

**Comment by sidolov** [comment]
Hi @bbmxxc. Thank you for your report. 
The issue has been fixed in magento/magento2#19195 by @eduard13 in 2.3-develop branch
Related commit(s):
 - [c9b52500a52067b8ccfc3329b7456512b6520526](https://github.com/magento/magento2/commit/c9b52500a52067b8ccfc3329b7456512b6520526)
 - [3b84172faf8612ca7fc54c4c3bb131923e521c01](https://github.com/magento/magento2/commit/3b84172faf8612ca7fc54c4c3bb131923e521c01)

The fix will be available with the upcoming 2.3.1 release.

**Comment by magento-engcom-team** [comment]
Hi @bbmxxc. Thank you for your report. 
The issue has been fixed in magento/magento2#19426 by @eduard13 in 2.2-develop branch
Related commit(s):
 - [b8fc2023e6b65604e1ffe73a2fb85a333176abc3](https://github.com/magento/magento2/commit/b8fc2023e6b65604e1ffe73a2fb85a333176abc3)

The fix will be available with the upcoming 2.2.8 release.

**Comment by wojtekn** [comment]
@magento-engcom-team  @eduard13 is it a proper fix? I didn't test it, but I see it sets current store on customer object. What if getById is used by CRON/CLI?

**Comment by Ctucker9233** [comment]
@magento-engcom-team The fix was not included in 2.2.8.

**Comment by ghost** [comment]
Hi @Ctucker9233 this issue has ben fixed by another commit, please always retest issue if you want to reopen ticket, thanks for collaboration https://github.com/magento/magento2/commit/d3a7b4007a9db4ac0a8b267f42540dd569b41461
